CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2BadStripConfigurableFakeESSource.cc

```diff
@@ -0,0 +1,255 @@
+public:
+  SiPhase2BadStripConfigurableFakeESSource(const edm::ParameterSet&);
+  ~SiPhase2BadStripConfigurableFakeESSource() override = default;
+
+  void setIntervalFor(const edm::eventsetup::EventSetupRecordKey&,
+                      const edm::IOVSyncValue& iov,
+                      edm::ValidityInterval& iValidity) override;
+
+  typedef std::unique_ptr<SiStripBadStrip> ReturnType;
+  ReturnType produce(const SiPhase2OuterTrackerBadStripRcd&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  std::map<unsigned short, unsigned short> clusterizeBadChannels(
+      const std::vector<Phase2TrackerDigi::PackedDigiType>& maskedChannels);
+
+  // configurables
+  bool m_printDebug;
+  float m_badComponentsFraction;
+
+  // random engine
+  std::unique_ptr<CLHEP::HepRandomEngine> m_engine;
+
+  // es tokens
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> m_trackTopoToken;
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> m_geomToken;
+};
+
+SiPhase2BadStripConfigurableFakeESSource::SiPhase2BadStripConfigurableFakeESSource(const edm::ParameterSet& iConfig)
+    : m_engine(new CLHEP::HepJamesRandom(iConfig.getParameter<unsigned int>("seed"))) {
+  auto cc = setWhatProduced(this);
+  m_trackTopoToken = cc.consumes();
+  m_geomToken = cc.consumes();
+
+  m_printDebug = iConfig.getUntrackedParameter<bool>("printDebug", false);
+  m_badComponentsFraction = iConfig.getParameter<double>("badComponentsFraction");
+
+  if (m_badComponentsFraction > 1. || m_badComponentsFraction < 0.) {
+    throw cms::Exception("Inconsistent configuration")
+        << "[SiPhase2BadStripChannelBuilder::c'tor] the requested fraction of bad components is unphysical. \n";
+  }
+
+  findingRecord<SiPhase2OuterTrackerBadStripRcd>();
+}
+
+void SiPhase2BadStripConfigurableFakeESSource::setIntervalFor(const edm::eventsetup::EventSetupRecordKey&,
+                                                              const edm::IOVSyncValue& iov,
+                                                              edm::ValidityInterval& iValidity) {
+  iValidity = edm::ValidityInterval{iov.beginOfTime(), iov.endOfTime()};
+}
+
+// ------------ method called to produce the data  ------------
+SiPhase2BadStripConfigurableFakeESSource::ReturnType SiPhase2BadStripConfigurableFakeESSource::produce(
+    const SiPhase2OuterTrackerBadStripRcd& iRecord) {
+  using namespace edm::es;
+  using Phase2TrackerGeomDetUnit = PixelGeomDetUnit;
+
+  //TrackerTopology const& tTopo = iRecord.get(m_trackTopoToken);
+  TrackerGeometry const& tGeom = iRecord.get(m_geomToken);
+
+  auto badStrips = std::make_unique<SiStripBadStrip>();
+
+  // early return with nullptr if fraction is ==0.
+  if (m_badComponentsFraction > 0.) {
+    ;  // do nothing
+  } else {
+    return badStrips;
+  }
+
+  LogDebug("SiPhase2BadStripConfigurableFakeESSource")
+      << " There are " << tGeom.detUnits().size() << " modules in this geometry." << std::endl;
```

Comment:
Just a minor comment here, is that you dont need `<< std::endl` (I know this is not very important and doesnt affect anything)

Reply:
applied in the last push.